drivers/usb/udc/udc_numaker.c

```diff
@@ -0,0 +1,1846 @@
+	NUMAKER_USBD_MSG_TYPE_SW_RECONN,
+};
+
+struct numaker_usbd_msg {
+	enum numaker_usbd_msg_type type;
+	union {
+		struct {
+			enum udc_event_type type;
+		} udc_bus_event;
+		struct {
+			uint8_t packet[8];
+		} setup;
+		struct {
+			uint8_t ep;
+		} out;
+		struct {
+			uint8_t ep;
+		} in;
+		struct {
+			uint8_t ep;
+		} xfer;
+	};
+};
+
+/* EP H/W context */
+struct numaker_usbd_ep {
+	bool valid;
+
+	const struct device *dev; /* Pointer to the containing device */
+
+	uint8_t ep_hw_idx;  /* BSP USBD driver EP index EP0, EP1, EP2, etc */
+	uint32_t ep_hw_cfg; /* BSP USBD driver EP configuration */
+
+	/* EP DMA buffer */
+	bool dmabuf_valid;
+	uint32_t dmabuf_base;
+	uint32_t dmabuf_size;
+
+	/* NOTE: On USBD, Setup and CTRL OUT are not completely separated. CTRL OUT MXPLD
+	 * can be overridden to 8 by next Setup. To overcome it, we make one copy of CTRL
+	 * OUT MXPLD immediately on its interrupt.
+	 */
+	uint32_t mxpld_ctrlout;
+
+	/* EP address */
+	bool addr_valid;
+	uint8_t addr;
+
+	/* EP MPS */
+	bool mps_valid;
+	uint16_t mps;
+};
+
+/* Immutable device context */
+struct udc_numaker_config {
+	struct udc_ep_config *ep_cfg_out;
+	struct udc_ep_config *ep_cfg_in;
+	uint32_t ep_cfg_out_size;
+	uint32_t ep_cfg_in_size;
+	USBD_T *base;
+	const struct reset_dt_spec reset;
+	uint32_t clk_modidx;
+	uint32_t clk_src;
+	uint32_t clk_div;
+	const struct device *clkctrl_dev;
+	void (*irq_config_func)(const struct device *dev);
+	void (*irq_unconfig_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t dmabuf_size;
+	bool disallow_iso_inout_same;
+	void (*make_thread)(const struct device *dev);
+};
+
+/* EP H/W context manager */
+struct numaker_usbd_ep_mgmt {
+	/* EP H/W context management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in udc_numaker_init().
+	 */
+	uint8_t ep_idx;
+
+	/* DMA buffer management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in udc_numaker_init().
+	 */
+	uint32_t dmabuf_pos;
+};
+
+/* Mutable device context */
+struct udc_numaker_data {
+	uint8_t addr; /* Host assigned USB device address */
+
+	struct k_msgq *msgq;
+
+	struct numaker_usbd_ep_mgmt ep_mgmt; /* EP management */
+
+	struct numaker_usbd_ep *ep_pool;
+	uint32_t ep_pool_size;
+
+	struct k_thread thread_data;
+};
+
+static inline void numaker_usbd_sw_connect(const struct device *dev)
+{
+	const struct udc_numaker_config *config = dev->config;
+	USBD_T *const base = config->base;
+
+	/* Clear all interrupts first for clean */
+	base->INTSTS = base->INTSTS;
+
+	/* Enable relevant interrupts */
+	base->INTEN = USBD_INT_BUS | USBD_INT_USB | USBD_INT_FLDET | USBD_INT_WAKEUP | USBD_INT_SOF;
+
+	/* Clear SE0 for connect */
+	base->SE0 &= ~USBD_DRVSE0;
+}
+
+static inline void numaker_usbd_sw_disconnect(const struct device *dev)
+{
+	const struct udc_numaker_config *config = dev->config;
+	USBD_T *const base = config->base;
+
+	/* Set SE0 for disconnect */
+	base->SE0 |= USBD_DRVSE0;
```

Comment:
This sounds wrong. Is the `SE0` simply some crazy naming to "disable D+ pull-up"? The device cannot initiate bus reset on its own - it is always host responsibility. The Full-Speed device can disable the D+ pull-up which will cause the pull-downs at the host to discharge the lines (and create SE0).

Is the NuMaker USBD publicly available? Does the HW allow disabling D+ pull-up?

Reply:
It is [publicly available](https://www.nuvoton.com/products/microcontrollers/arm-cortex-m4-mcus/m467-ethernet-crypto-series/?group=Document&tab=2) (check the title M463/M467 Series Technical Reference Manual and the IP is USB 1.1 Device Controller (USBD)).

According to the h/w spec, the effect is to force D+/D- to low, so it can simulate disconnecting device from host. Excerpt the description:

> A software-disconnect function is also supported for this USB controller. It is used to simulate the disconnection of this device from the host. If user enables SE0 bit (USBD_SE0), the USB controller will force the output of USB_D+ and USB_D- to level low and its function is disabled. After disabling the SE0 bit, host will enumerate the USB device again.

Reply:
I wonder if setting `USBD_SE0` is really standard compliant. The specification describes the disconnection as D+ pull-up removal. Anyway, device doesn't have option to disable the pull-up so this is what the software has to do.